gadget/update.go

```diff
@@ -1197,21 +1170,58 @@ func volumeStructureToLocationMapImpl(old GadgetData, mod Model, vols map[string
 	// location to update given the VolumeStructure
 	return buildVolumeStructureToLocation(
 		mod,
-		old,
-		vols,
+		oldVolumes,
+		newVolumes,
 		volToDeviceMapping,
 		missingInitialMapping,
 	)
 }
 
+func validateVolumesMatch(old, new map[string]*Volume) error {
+	oldVolumes := make([]string, 0, len(old))
+	newVolumes := make([]string, 0, len(new))
+
+	for oldVol := range old {
+		oldVolumes = append(oldVolumes, oldVol)
+	}
+	for newVol := range new {
+		newVolumes = append(newVolumes, newVol)
+	}
+	common := strutil.Intersection(newVolumes, oldVolumes)
+	// check dissimilar cases between common, new and old
+	switch {
+	case len(common) != len(newVolumes) && len(common) != len(oldVolumes):
+		// there are both volumes removed from old and volumes added to new
+		return fmt.Errorf("cannot update gadget assets: volumes were both added and removed")
+	case len(common) != len(newVolumes):
+		// then there are volumes in old that are not in new, i.e. a volume
+		// was removed
+		return fmt.Errorf("cannot update gadget assets: volumes were removed")
+	case len(common) != len(oldVolumes):
+		// then there are volumes in new that are not in old, i.e. a volume
+		// was added
+		return fmt.Errorf("cannot update gadget assets: volumes were added")
+	}
```

Comment:
maybe we could use it as an opportunity to simplify, but since it's just moving preexeisting code it's ok to leave it as it is:
```suggestion
	if len(common) != len(newVolumes) || len(common) != len(oldVolumes) {
		// there are both volumes removed from old and volumes added to new
		return fmt.Errorf("cannot update gadget assets: differing volumes in new and old gadget")
	}
```

Reply:
i left this as an exercise for a separate PR